text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,648 @@
+
+If any of these checks fail a compiler error is issued.
+
+One additional check must be done, if the backend cannot guarantee that TCE will be performed an ICE is issued. It is also suggested to ensure that the invariants provided by the pre-requisites are maintained during compilation, raising an ICE if this is not the case.
+
+Note that as `become` is a keyword reserved for exactly the use-case described in this RFC there is no backward-compatibility break.
+
+This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.
+
+See below for the reasoning why operators are not supported.
+
+## Operators are not supported
+Invocations of operators were considered as valid targets but were rejected on grounds of being too error-prone.
+In any case, these can still be called as methods. One example of their error-prone nature ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1167112296)):
+```rust
+pub fn fibonacci(n: u64) -> u64 {
+    if n < 2 {
+        return n
+    }
+    become fibonacci(n - 2) + fibonacci(n - 1)
+}
+```
+In this case, a naive author might assume that this is going to be a stack space-efficient implementation since it uses tail recursion instead of normal recursion. However, the outcome is more or less the same since the critical recursive calls are not actually in tail call position.
+
+Further confusion could result from the same-signature restriction where the Rust compiler raises an error since fibonacci and <u64 as Add>::add do not share a common signature.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+<!-- Why should we *not* do this? -->
+As this feature should be mostly independent of other features the main drawback lies in the implementation and
+maintenance effort. This feature adds a new keyword which will need to be implemented not only in Rust but also in other
+tooling. The primary effort, however, lies in supporting this feature in the backends:
+- LLVM supports a `musttail` marker to indicate that TCE should be performed [docs](https://llvm.org/docs/LangRef.html#id327). Clang which already depends on this feature seems to only generate correct code for the x86 backend [source](https://github.com/rust-lang/rfcs/issues/2691#issuecomment-1490009983) (as of 30.03.23).
+- GCC does seem to support an equivalent `musttail` marker, though it is only accessible via the [libgccjit API](https://gcc.gnu.org/onlinedocs/gcc-7.3.0/jit/topics/expressions.html#gcc_jit_rvalue_set_bool_require_tail_call) ([source](https://github.com/rust-lang/rfcs/pull/3407#discussion_r1160013809)).
+- WebAssembly accepted tail calls into the [standard](https://github.com/WebAssembly/proposals/pull/157/) and Cranelift is now [working](https://github.com/bytecodealliance/rfcs/pull/29) towards supporting it.
```

Comment:
None of the wasm targets enable the tail call feature and they likely can't enable it for a long time due to back compat reasons.

Reply:
Also Cranelift will only support tail calls for the ABI used for Webassembly. It will not support it for the SystemV or WindowsFastcall ABI's used for rust code.

Reply:
> None of the wasm targets enable the tail call feature and they likely can't enable it for a long time due to back compat reasons.

As of now `tail-calls` are not yet part of the WebAssembly standard so I'd assume that there will at least be the option to generate them even though this option will probably be disabled by default. That's the same with other post-MVP Wasm proposals from what I know. @bjorn3 Please correct me if I wrong about this assumption?

> Also Cranelift will only support tail calls for the ABI used for Webassembly. It will not support it for the SystemV or WindowsFastcall ABI's used for rust code.

Does this mean that Rust code needs something like `extern "Wasm"` applied to functions for tail calls to work under Cranelift? I am confused. 🤔

Your points should not be a blocker for this RFC to be accepted but they might be blocker for the stabilization of it later on.

Reply:
> As of now tail-calls are not yet part of the WebAssembly standard but proposed to be so I'd assume that there will at least be the option to generate them for producers even though this option will probably be disabled by default. That's the same with other post-MVP Wasm proposals from what I know. @bjorn3 Please correct me if I wrong about this assumption?

Yeah, I did assume LLVM has a target feature to enable it, however rustc can't enable this target feature for wasm32-unknown-unknown as it did break running rustc compiled wasm modules on wasm runtimes that don't support tail calls. This likely happens even if the user didn't explicitly wrote a tail call as LLVM will likely optimize some calls into tail calls.

> Does this mean that Rust code needs something like extern "Wasm" applied to functions for tail calls to work under Cranelift? I am confused. thinking

Cranelift supports multiple calling conventions like SystemV and WindowsFastcall. These are the default calling conventions for C on Unix and Windows respectively. The Rust ABI used by the LLVM backend is based on the same calling convention as used by C. To remain ABI compatible, cg_clif uses the same calling convention as the LLVM backend. Cranelift will not support tail calls for the SystemV and WindowsFastcall calling conventions. Instead it will only support it for the Cranelift specific Tail calling convention designed specifically for use in Webassembly runtimes. This calling convention is not supported by LLVM, so we can neither switch the default Rust ABI to it, not expose it as `extern "Tail"` in Rust. In other words cg_clif can't remain ABI compatible with cg_llvm and support explicit tail calls. Furthermore the Tail calling convention exclusively uses the stack for passing arguments, making it slower, but also much easier to implement tail calls.

Reply:
Thank you for the info @bjorn3, I'll add a link to this discussion in the RFC.

Reply:
For the signature matching case we can do defunctionalization and TCE in MIR level.

Reply:
Based on https://github.com/bytecodealliance/wasmtime/issues/1065#issuecomment-1367015529 I think it isn't possible to implement in the general case for the x86_64 SystemV calling convention.

> The general problem to wrestle with is easiest to see (IMHO) in a minimized example: f and g tail-call each other, both have enough arguments that some of them are passed on the stack, and they have different numbers of arguments. E.g., on x86-64 where the first six i64 args are passed in registers, say that f has 8 i64 args and g has 10 i64 args.
>
> Then when f tail-calls g, it needs 32 bytes of stack-arg space (the remaining 4 i64 args), but it itself only was given 16 bytes of stack-arg space. Furthermore, the original caller of f, in the standard SysV ABI, is responsible for popping those 16 bytes of space. There's no way to satisfy all the constraints for a true tail-call in this case in the SysV ABI. That's why the usual way to achieve true tail-calls in the arbitrary case is to adopt a "callee-pop" ABI instead: f and g both expect a certain number of args on the stack, and pop them before returning. This way, f's tail-call to g can do that pop, and then push g's args; and when g returns to the original caller of f, there's no need to somehow know that the actual returnee had 32 bytes rather than 16 of stack args.

@cfallin Do I understand this correctly?

Reply:
If you restrict tail calls to functions with the exact same signature then all you need to do is overwrite the arguments of the current function in-place with new values.

Reply:
> Based on [bytecodealliance/wasmtime#1065 (comment)](https://github.com/bytecodealliance/wasmtime/issues/1065#issuecomment-1367015529) I think it isn't possible to implement in the general case for the x86_64 SystemV calling convention.
> 
> > The general problem to wrestle with is easiest to see (IMHO) in a minimized example: f and g tail-call each other, both have enough arguments that some of them are passed on the stack, and they have different numbers of arguments. E.g., on x86-64 where the first six i64 args are passed in registers, say that f has 8 i64 args and g has 10 i64 args.
> > Then when f tail-calls g, it needs 32 bytes of stack-arg space (the remaining 4 i64 args), but it itself only was given 16 bytes of stack-arg space. Furthermore, the original caller of f, in the standard SysV ABI, is responsible for popping those 16 bytes of space. There's no way to satisfy all the constraints for a true tail-call in this case in the SysV ABI. That's why the usual way to achieve true tail-calls in the arbitrary case is to adopt a "callee-pop" ABI instead: f and g both expect a certain number of args on the stack, and pop them before returning. This way, f's tail-call to g can do that pop, and then push g's args; and when g returns to the original caller of f, there's no need to somehow know that the actual returnee had 32 bytes rather than 16 of stack args.
> 
> @cfallin Do I understand this correctly?

This can be worked around by using a trampoline.  f and g gain an additional boolean argument that indicates if they were called by the trampoline or not.  If they were, then true is passed, and tail call is performed by returning to the trampoline, which then pops everything and makes the next call.  Non-tail calls are made to a wrapper that passes false.  The extra parameter can be stored in a caller-save scratch register on ABIs that have that.

Reply:
> That said, the extern "Rust" ABI is explicitly unspecified, so Rust could choose to use a callee-pops calling convention to make arbitrary tail calls fast.

There are several calling conventions in LLVM that support tail calls, but none of them are usable for rustc's purpose on all platforms:

* "fastcc"/"tailcc": AFAIK this one is not stable across LLVM version, which is required to keep other backends ABI compatible. I'm also not sure if it is allowed for functions exported from the current cgu.
* "cc 10" (GHC call conv): Only supports x86 and x86_64
* "cc 11" (HiPE call conv): Only supports x86 and x86_64
* "swifttailcc": x86/x86_64 and aarch64 only I think

Also the only targets on which LLVM supports tail calls are x86/x86_64 (only linkage unit local in case of PIC code like we do by default), PowerPC (only linkage unit local functions), AArch64, and WebAssembly.

Reply:
As for the trampoline option, that would work I guess, but is slower and would complicate all backends to support it.

Reply:
> This can be worked around by using a trampoline. f and g gain an additional boolean argument that indicates if they were called by the trampoline or not. If they were, then true is passed, and tail call is performed by returning to the trampoline, which then pops everything and makes the next call. Non-tail calls are made to a wrapper that passes false. The extra parameter can be stored in a caller-save scratch register on ABIs that have that.

Indeed, but now you've invented a calling convention that isn't quite SysV. The appeal of a "true" SysV tail-call solution is that a caller can call a callee and *not* set up this extra parameter, and have no knowledge of the tail-call convention, and everything would still work.

I'm also generally not too fond of trampoline-based approaches: they exert a pressure on users of tail calls and discourage their use in general. If we're going to have tail calls, it's preferable that they're as efficient or almost as efficient as ordinary calls -- that will unlock the truly interesting use-cases (languages unconditionally using tail-call instructions in tail positions to give Scheme-like guarantees; compiling every basic block to a little function and tail-calling to work around structured control flow limitations; using tail calls for ICs (inline caches); and maybe others).

Reply:
> If we're going to have tail calls, it's preferable that they're as efficient or almost as efficient as ordinary calls

I agree, but that will require a callee-pops calling convention.

Reply:
@nbdd0121 correct me if I'm wrong but defunctionalization will only work for statically known function calls, not dynamic calls or calls to external functions.

@bjorn3 and @cfallin what are your thoughts on supporting tail calls only between functions with matching function signatures? As @Amanieu said, supporting tail calls with this restriction should be possible independent of the calling convention.

Reply:
@phi-go I think supporting this restriction at least is still a very nice and usable building-block!

Reply:
Updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/b6b309492602bf0a5b1733022234110ca2678632

Let me know if you disagree with the wording or want another change.